crates/bevy_text/src/pipeline.rs

```diff
@@ -1,52 +1,147 @@
+    map_handle_to_font_id: HashMap<AssetId<Font>, cosmic_text::fontdb::ID>,
+    /// The font system is used to retrieve fonts and their information, including glyph outlines.
+    ///
+    /// See [`cosmic_text::FontSystem`] for more information.
+    font_system: FontSystem,
+    /// The swash cache rasterizer is used to rasterize glyphs
+    ///
+    /// See [`cosmic_text::SwashCache`] for more information.
+    swash_cache: SwashCache,
 }
 
 impl TextPipeline {
-    pub fn get_or_insert_font_id(&mut self, handle: &Handle<Font>, font: &Font) -> FontId {
-        let brush = &mut self.brush;
-        *self
-            .map_font_id
-            .entry(handle.id())
-            .or_insert_with(|| brush.add_font(handle.id(), font.font.clone()))
+    pub fn create_buffer(
+        &mut self,
+        fonts: &Assets<Font>,
+        sections: &[TextSection],
+        linebreak_behavior: BreakLineOn,
+        bounds: Vec2,
+        scale_factor: f64,
+    ) -> Result<Buffer, TextError> {
+        // TODO: Support multiple section font sizes, pending upstream implementation in cosmic_text
+        // For now, just use the first section's size or a default
+        let font_size = sections
+            .get(0)
+            .map(|s| s.style.font_size)
+            .unwrap_or_else(|| crate::TextStyle::default().font_size)
+            as f64
+            * scale_factor;
+        // TODO: Support line height as an option. Unitless `1.2` is the default used in browsers (1.2x font size).
+        let line_height = font_size * 1.2;
+        let (font_size, line_height) = (font_size as f32, line_height as f32);
+        let metrics = Metrics::new(font_size, line_height);
+
+        let font_system = &mut acquire_font_system(&mut self.font_system)?;
+
+        let spans: Vec<(&str, Attrs)> = sections
+            .iter()
+            .enumerate()
+            .map(|(section_index, section)| {
+                (
+                    &section.value[..],
+                    get_attrs(
+                        section,
+                        section_index,
+                        font_system,
+                        &mut self.map_handle_to_font_id,
+                        fonts,
+                    ),
+                )
+            })
+            .collect();
+
+        // TODO: cache buffers (see Iced / glyphon)
+        let mut buffer = Buffer::new_empty(metrics);
+        let buffer_height = f32::INFINITY;
+        buffer.set_size(font_system, bounds.x.ceil(), buffer_height);
+
+        buffer.set_wrap(
+            font_system,
+            match linebreak_behavior {
+                BreakLineOn::WordBoundary => Wrap::Word,
+                BreakLineOn::AnyCharacter => Wrap::Glyph,
+                BreakLineOn::NoWrap => Wrap::None,
+            },
+        );
+
+        // TODO: other shaping methods?
+        buffer.set_rich_text(font_system, spans, Shaping::Basic);
```

Comment:
`Shaping::Basic` won't load system fonts, so emoji won't be displayed unless emoji font is loaded explicitly😢

Reply:
Yeah, system font support will be handled in another PR, and then smileys can work :)